gate-deck-plugins/src/main/kotlin/com/netflix/spinnaker/gate/plugins/DeckPluginCache.kt

```diff
@@ -0,0 +1,145 @@
+ * limitations under the License.
+ */
+package com.netflix.spinnaker.gate.plugins
+
+import com.netflix.spectator.api.Id
+import com.netflix.spectator.api.Registry
+import com.netflix.spinnaker.kork.plugins.bundle.PluginBundleExtractor
+import com.netflix.spinnaker.kork.plugins.update.PluginUpdateService
+import com.netflix.spinnaker.kork.plugins.update.SpinnakerUpdateManager
+import org.pf4j.update.PluginInfo
+import org.slf4j.LoggerFactory
+import org.springframework.scheduling.annotation.Scheduled
+import org.springframework.stereotype.Component
+import java.nio.file.Files
+import java.nio.file.Path
+import java.nio.file.StandardCopyOption
+
+/**
+ * Responsible for keeping an up-to-date cache of all plugins that Deck needs to know about.
+ *
+ * TODO(rz): Currently downloads all plugins and inspects the extracted files to determine if a plugin has deck assets.
+ *  This is pretty inefficient - would be better to have a metadata file shipped alongside a bundle.
+ * TODO(rz): Downloads all plugins to an isolated temp directory, so if Gate itself has plugins, this cache will be
+ *  downloading an intersection of plugins twice.
+ */
+@Component
+class DeckPluginCache(
+  private val updateManager: SpinnakerUpdateManager,
+  private val updateService: PluginUpdateService,
+  private val pluginBundleExtractor: PluginBundleExtractor,
+  private val registry: Registry
+) {
+
+  private val log by lazy { LoggerFactory.getLogger(javaClass) }
+
+  private val cache: MutableSet<PluginCacheEntry> = mutableSetOf()
+  private var cachePopulated: Boolean = false
+
+  private val versionsId = registry.createId("plugins.deckCache.versions")
+  private val hitsId = registry.createId("plugins.deckCache.hits")
+  private val missesId = registry.createId("plugins.deckCache.misses")
+  private val downloadDurationId = registry.createId("plugins.deckCache.downloadDuration")
+  private val refreshDurationId = registry.createId("plugins.deckCache.refreshDuration")
+
+  /**
+   * Refreshes the local file cache of _current_ plugins. Should Deck need plugin assets from an older plugin release
+   * version, it will be downloaded and cached on-demand.
+   *
+   * The default refresh interval is 5 minutes.
+   */
+  @Scheduled(fixedDelayString = "\${spinnaker.extensibility.deck.cache-refresh-interval-ms:300000}", initialDelay = 0)
+  internal fun refresh() {
+    registry.timer(refreshDurationId).record {
+      log.info("Refreshing plugin cache")
+
+      updateManager.refresh()
+
+      val newCache = updateManager.plugins
+        .map {
+          val plugin = DeckPluginVersion(it.id, selectPluginVersion(it))
+          val path = getOrDownload(plugin.id, plugin.version)
+          PluginCacheEntry(plugin, path)
+        }
+        .filter {
+          // TODO(rz): Once bundles support manifest artifacts, we can just inspect those
+          //  for a deck plugin rather than downloading everything and seeing if a
+          //  index.js file exists.
```

Comment:
I'm not so sure about this.  What do you mean by manifest artifacts?  I think if we bundle plugins, that's one artifact, so we still need to download the entire artifact and then determine which plugin inside the bundle is for Deck.  Of course, we shouldn't ever need to download a bundle that does not contain a Deck plugin, but I think the mechanisms for that are already in place.

Reply:
The thought from the plugin sync was that we could produce the `bundle.zip` but also create a manifest file, like a `.pom` that would describe what is inside the bundle without having to download, extract and inspect it. The metadata for determining what is inside a bundle doesn't exist yet (or we need to change how `requires` works)

Reply:
Yeah, being able to access manifests would be great.  I'd love to get to a place where all the plugins' manifests can be refreshed frequently and in a light-weight manner.